### PR TITLE
Implement annotations for Services and Content Providers

### LINKED ISSUE #679: Support access to the manifest file for AI Extensions
Extensions can already define UserLibraries and Permissions, but there's no way to create extensions that need other type of access to the manifest through Compiler.java

A first iteration of this could be to move the BroadcastReceiver code used in Texting.java to the component (with an annotation) instead of adding it directly in Compiler.java. After that, more support for things like Services, minSDK or finer grain permissions should be added.

### PR DESCRIPTION
**Pull request by ShreyashSaitwal**
This PR adds support for including `<service>` and `<provider>` tags in Android Manifest by using the `@UsesServices` and `@UsesContentProviders` annotations respectively. Also, it resolves #679.

@ewpatton @jisqyv @josmas please review.

### CODE REVIEW

**Comment by AppInventorWorkerBee** [comment]
Can one of the admins verify this patch?

**Comment by ewpatton** [comment]
@AppInventorWorkerBee ok to test

**Comment by ShreyashSaitwal** [comment]
@ewpatton 
Ain't that label should to be `affects: ucr` instead of `affects: master`?

**Comment by ShreyashSaitwal** [comment]
Hi @ewpatton,
Are there any chances of this change being staged for nb184 release?

**Comment by ewpatton** [comment]
> Hi @ewpatton,
> Are there any chances of this change being staged for nb184 release?

Yes. I tested it by making some simple extensions with services to try it out. Very nice work, I think it's going to open a lot of possibilities in extensions. If you can address the two issues I pointed out I think we can go ahead and merge this.

**Comment by halatmit** [comment]
When this is released would it be possible up update the documentation on App Inventor Extensions
https://docs.google.com/document/d/1orlYWGlHOY0hmAC0zjadQ4DET2K8U34C8hv5zmuxZLU/edit#heading=h.q0qxs8avzosb

**Comment by ewpatton** [comment]
@halatmit Yes. I also have a separate document I've been working on that documents the annotations in more detail. I've been updating that based on the changes in this PR.

File: appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java [hunk 1]
```diff
@@ -600,6 +616,48 @@ void generateBroadcastReceivers() {
...
   }
 
+  /*
+   * Generate a set of conditionally included services needed by this project.
+   */
+  @VisibleForTesting
+  void generateServices() {
+    try {
+      loadJsonInfo(servicesNeeded, ComponentDescriptorConstants.SERVICES_TARGET);
+    }
```
[the remark this answers is not in the document]
Comment (ShreyashSaitwal) [reply appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java]: Fixed. Thanks for pointing that out.

File: appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java [hunk 2]
```diff
@@ -600,6 +616,48 @@ void generateBroadcastReceivers() {
...
+  }
+
+  /*
+   * Generate a set of conditionally included content providers needed by this project.
+   */
+  @VisibleForTesting
+  void generateContentProviders() {
+    try {
+      loadJsonInfo(contentProvidersNeeded, ComponentDescriptorConstants.CONTENT_PROVIDERS_TARGET);
+    }
```
[the remark this answers is not in the document]
Comment (ShreyashSaitwal) [reply appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java]: Fixed. Thanks for pointing that out.

**Comment by ShreyashSaitwal** [comment]
> Two minor code styles issues but otherwise LGTM.

Hi @ewpatton,
I've updated the code style.

**Pull request merged**